Support for barge-ins and message repeats.

rudigiesler opened the pull request:
We'd like to add support for people to respond mid-message or mid-message sequence and for messages to repeat if there is no response after some period.

Use cases:

* A person knows the option they would like to choose and selects it mid playback. We'd like to have the message playback end and for the selection to be returned to the application for processing.
* A person fails to respond and the messages is repeated to them (a configurable number of times before the transport hangs up the call and notifies the application).
* A person explicitly responds with a configurable DTMF command that asks for the message to be repeated (e.g. `*#` during a message causes the message to be repeated).

Freeswitch features that might help us implement this:

* `play-and-get-digits` (https://freeswitch.org/confluence/display/FREESWITCH/mod_dptools%3A+play+and+get+digits)
* `play-and-detect-speech` (https://freeswitch.org/confluence/display/FREESWITCH/mod_dptools%3A+play_and_detect_speech)
* `play`, with `playback_terminators` (https://freeswitch.org/confluence/display/FREESWITCH/mod_dptools%3A+playback)
* `flush_dtmf` (https://freeswitch.org/confluence/display/FREESWITCH/mod_dptools%3A+flush+dtmf)

rudigiesler commented:
Ready for review.

gsvr commented on README.rst:
```diff
@@ -23,3 +23,52 @@ You can contact the Vumi development team in the following ways:
+    A boolean value that if ``True``, stops the playback of the message when
+    a DTMF character arrives. This allows the response to the input to be
+    played immediately, rather than waiting for the first message to finish
+    playing before hearing the response message.
```
@rudigiesler What is the default value for `barge_in`?

rudigiesler replied:
There is no default value for `barge_in`, which makes the value `None`, which evaluates the same as `False` would. Maybe we should mention that the default value is `False`?

hodgestar replied:
Agreed -- let's mention the default, just to clear.

hodgestar commented on README.rst:
```diff
@@ -23,3 +23,52 @@ You can contact the Vumi development team in the following ways:
+-----
+
+Voice transports may receive additional hints for how to handle outbound
+messages the ``voice`` section of ``helper_metadata``. The ``voice`` section
```
s/the/in the/

hodgestar commented on README.rst:
```diff
@@ -23,3 +23,52 @@ You can contact the Vumi development team in the following ways:
+    played immediately, rather than waiting for the first message to finish
+    playing before hearing the response message.
+:``tries``:
+   If ``barge_in`` is ``True``, this will set the amount of times a message is
```
s/amount/number/

hodgestar commented on README.rst:
```diff
@@ -23,3 +23,52 @@ You can contact the Vumi development team in the following ways:
+   played if no input is received. Defaults to ``1``.
+:``time_gap``:
+   If ``barge_in`` is ``True`` and ``tries`` is greater than ``1``, this
+   specifies the time gap (in ms) that is given before repeating the message,
```
I would maybe write `specifies the length of the pause (in ms) between repeats of the message`.

hodgestar commented on vxfreeswitch/voice.py:
```diff
@@ -92,25 +94,25 @@ def create_and_stream_text_as_speech(self, folder, command, ext, message):
         yield self.set("tts_engine=" + engine)
         yield self.set("tts_voice=" + voice)
-        yield self.execute("speak", message)
+        yield self.output_stream("say:'%s'" % message, settings)
```
Do we need to escape single quotes in `message` somehow?

hodgestar replied:
I'm guessing that this is intended to be a URL schema, but the details are likely custom to FreeSwitch?

hodgestar replied:
The FreeSwitch docs (https://wiki.freeswitch.org/wiki/Misc._Dialplan_Tools_say) are making me worry that `say` is rather different to `speak`. Is there a difference in practice? If so, perhaps it's possible to use `speak` here anyway? Not sure.

rudigiesler replied:
The docs are a bit confusing here. I have tried to use `speak` here, but it doesn't work. Using `say` here works exactly the same as `speak` in every test that I've tried. I got the idea from the example in this doc: https://freeswitch.org/confluence/display/FREESWITCH/mod_dptools:+play+and+get+digits

Even when I try to put a `say` command like `en name_spelled iterated ab733#`, it acts as if I'm passing a string to `speak`.

rudigiesler replied:
We do need to escape the single quotes though.

hodgestar commented on vxfreeswitch/voice.py:
```diff
@@ -118,13 +120,32 @@ def stream_text_as_speech(self, message):
-        self.log("Playing back URL: %r" % (url,))
-        return self.playback(url)
+    def output_message(self, text, settings={}):
+        return self.stream_text_as_speech(text, settings=settings)
```
Did we intentionally drop the `self.log("Sending text as speech: %r" % (text,))` here?

hodgestar replied:
Ah, I see it's replaced by log in `output_stream`. Please ignore previous comment.

hodgestar commented on vxfreeswitch/voice.py:
```diff
@@ -118,13 +120,32 @@ def stream_text_as_speech(self, message):
+        self.log("Playing back: %r" % (message,))
+        if settings.get('barge_in'):
+            terminator = settings.get('wait_for')
+            if settings.get('wait_for') is None:
```
This could be just `if terminator is None:`.

hodgestar commented on vxfreeswitch/voice.py:
```diff
@@ -118,13 +120,32 @@ def stream_text_as_speech(self, message):
+            if settings.get('wait_for') is None:
+                # We just want to get 1 digit
+                minimum, maximum = 1, 1
+                terminator = "' '"
```
What setting ` ` as the terminator do? Maybe worth a comment?

rudigiesler replied:
It makes it as if there is no terminator character. This was back from when when I was using the result from `play_and_get_digits` rather than the existing DTMF things that we have, and having a terminator character meant that if they pressed the terminator character, the result would be an empty string.

This could also be the empty string `"''"`. I'll change it to that and add a comment to make it clearer.

hodgestar commented on vxfreeswitch/voice.py:
```diff
@@ -118,13 +120,32 @@ def stream_text_as_speech(self, message):
+            invalid_message = 'silence_stream://1'
+            return self.execute('play_and_get_digits', ' '.join([
+                str(minimum), str(maximum), str(tries), str(timeout),
+                str(terminator), message, invalid_message, self.VAR_NAME,
```
We set `VAR_NAME = 'digits'` but use it only here. What does it do?

rudigiesler replied:
This was also back when I got the result from the execute complete instead of through the DTMF things that we already have. This is where freeswitch stores the result when `play_and_get_digits` completes.

rudigiesler replied:
I'll remove this since we don't need it anymore.

hodgestar commented on vxfreeswitch/voice.py:
```diff
@@ -392,12 +420,10 @@ def send_outbound_message(self, client, message):
 
         overrideURL = None
         client.set_input_type(None)
```
I think we can delete the two lines above now since `overrideURL` and `input_type` are always set below?

hodgestar commented:
Left some questions, but really happy with how clean this looks!

rudigiesler commented:
@hodgestar I think I've addressed all of your comments. Ready for another review. I was also worried about the `say:` command by looking at the docs, but the examples seem to hint that it works the same as `speak`, and in all my testing I can't find a difference between `speak` and the `say:` command.

hodgestar commented:
@rudigiesler Yeah, the FreeSwitch docs are a bit crazy. Can we:
* add a test for the single quote escaping
* leave a comment in the code about the FreeSwitch `say` documentation being misleading and that `say` in this context functions identically to the `speak` command?
Otherwise looks good!

rudigiesler commented:
@hodgestar I've addressed your last comments. Ready for a (hopefully) final review?

hodgestar commented:
:+1: once Travis builds are done.
